lib/index.ts

```diff
@@ -27,26 +44,161 @@ export function inspect(root, targetFile, options) {
+  };
+}
+
+// Applies upgrades to direct and indirect dependencies
+export async function applyRemediationToManifests(
+  root: string,
+  manifests: ManifestFiles,
+  upgrades: DependencyUpdates,
+  options: Options
+) {
+  const manifestNames = Object.keys(manifests);
+  const targetFile = manifestNames.find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  );
+  if (
+    !targetFile ||
+    !manifestNames.every(
+      (fn) =>
+        path.basename(fn) === 'requirements.txt' ||
+        path.basename(fn) === 'constraints.txt'
+    )
+  ) {
+    throw new Error(
+      'Remediation only supported for requirements.txt and constraints.txt files'
+    );
+  }
+
+  const provOptions = { ...options };
+  provOptions.args = provOptions.args || [];
+  provOptions.args.push('--only-provenance');
+
+  const topLevelDeps = (await inspect(root, targetFile, provOptions)).package;
+  applyUpgrades(manifests, upgrades, topLevelDeps);
+
+  return manifests;
+}
+
+function applyUpgrades(
+  manifests: ManifestFiles,
+  upgrades: DependencyUpdates,
+  topLevelDeps: DepTree
+) {
+  const requirementsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'requirements.txt'
+  ) as string;
+  const constraintsFileName = Object.keys(manifests).find(
+    (fn) => path.basename(fn) === 'constraints.txt'
+  );
+
+  // Updates to requirements.txt
+  const patch: { [zeroBasedIndex: number]: string | false } = {}; // false means remove the line
+  const append: string[] = [];
+
+  for (const upgradeFrom of Object.keys(upgrades)) {
+    const pkgName = upgradeFrom.split('@')[0].toLowerCase();
+    const newVersion = upgrades[upgradeFrom].upgradeTo.split('@')[1];
+    const topLevelDep = (topLevelDeps.dependencies || {})[
+      pkgName
+    ] as DepTreeDep;
+    if (topLevelDep && topLevelDep.labels && topLevelDep.labels.provenance) {
+      const lineNubers = topLevelDep.labels.provenance
+        .split(':')[1]
+        .split('-')
+        .map((x) => parseInt(x));
+      // TODO: what if the original version spec was limited, e.g. >=1.0,<2.0 ?
+      patch[lineNubers[0] - 1] = pkgName + '>=' + newVersion;
+      if (lineNubers.length >= 1) {
+        for (let i = lineNubers[0] + 1; i <= lineNubers[1]; i++) {
+          patch[i - 1] = false;
+        }
+      }
+    } else {
+      // The dependency is not a top level: we are pinning a transitive using constraints file.
+      if (!constraintsFileName) {
+        append.push(
+          pkgName +
+            '>=' +
+            newVersion +
+            ' # not directly required, pinned by Snyk to avoid a vulnerability'
+        );
+      } else {
+        // TODO: parse constraints and replace the pre-existing one if any
+        const lines = manifests[constraintsFileName].trim().split('\n');
+        lines.push(
+          pkgName +
+            '>=' +
+            newVersion +
+            ' # pinned by Snyk to avoid a vulnerability'
```

Comment:
We should definitely elaborate on this; the vuln id for example, but not relevant for now